### ISSUE TITLE
Random rollbacks

### ISSUE THREAD

**Issue by BbFGE** [Microsoft/cordova-plugin-code-push#356]
Phonegap Build
I have been using Phonegap Build to build both Android (apk) and iOS (ipa) files.
The documentation for Codepush (https://docs.microsoft.com/en-us/appcenter/distribution/codepush/cordova) states that the plugin is compatible with Phonegap Build. However, I have built apps for both iOS and Android (adding <preference name="android-build-tool" value="gradle" /> to the config.xml as recommended for Android) however, neither app would pick up new releases from Codepush - the app would simply display as per the original installation.

Random rollbacks
I have been having major setbacks testing Codepush with Android. I install the compiled app on a device, and it runs ok. I amend the app by making a minor change to the html file and use appcenter codepush release-cordova... to push out a new update.

I then run the app, close it and run it again. The first two or 3 updates will be correctly picked up. However, after this (and it is not consistent, it differs between devices when this happens), the next update fails and the app reverts to the installed version and in appcenter, the update is shown under 'Rollbacks'.

From that point, any new releases are ignored and the app stays on the installed version.
The only two fixes seem to be
1. Re-installing the app - this fixes it and it starts picking up new releases, until it fails again OR
2. Clearing the entire history for the deployment. Again, this fixes it until it fails again.

The update isn't crashing - I have merely changed some text in the index.html so it wouldn't crash (plus an update will install on one device but fail on another so it can't be that).

The app triggers the Codepush sync as follows : 
document.addEventListener("deviceready", onDeviceReady, false);

function onDeviceReady() {
	codePush.sync(null,{
		ignoreFailedUpdates : false
	});
}
I have tried it with and without the 'ignoreFailedUpdates' option, but it fails eventually either way.
I assume the problem is caused by the app failing to complete codePush.notifyApplicationReady, perhaps due to network problems. 

However, this seems to be a show-stopper in that it effectively makes Codepush unusable in the real world. There is no guarantee updates will reach users; worse still, if the update fails, the app rolls back to the last installed version and subsequently ignores all new updates, so they are stuck at an old version or have to go through the pain of re-installing the app. (Clearing the history for the deployment every time we do an update might work, but no guarantees).

As an example, I created an app that simly displayed some html, including a version number.
Build and installed 
'v1.00'
edited the html, changing the version to v1.01.
Release the update (having already set the app using appcenter apps set-current -a ...)
appcenter codepush release-cordova -d Staging
Open app, close it, open again
Correctly shows
'v1.01'
edited the html, changing the version to v1.02.
Release the update 
Open app, correctly shows
'v1.02'
Close it, then re-open it
Shows 'v1.00', and appcenter shows rollbacks.
I have tried using the '-m' option on appcenter codepush release-cordova
Also, I have tried it on both Android 6 and 7 with the same problems.

My questions are:
1. Is there anything else I can try to get the updates to install correctly, or, if they fail, NOT to freeze the app at its last installed state and ignore any future updates.
2. Is this a cordova problem or is the same thing likely to happen if I switched using React-Native?
3. Is  there a way to make Codepush work with Phonegap Build?#

We wanted to be able to roll out an app to our user base but without having to have them re-install on every small change. Codepush seemed a perfect solution, however, at the moment, is it is looking as if it simply isn't workable.

Any help would be much appreciated.

Info from config.xml

  <content src="index.html" />
    <access origin="*" />
    <platform name="android">
        <preference name="CodePushDeploymentKey" value="...." />
        <preference name="APP_SECRET" value="..." />
    </platform>
    <plugin name="cordova-plugin-code-push" spec="~1.11.2" />
    <plugin name="cordova-plugin-whitelist" spec="~1.3.3" />
    <plugin name="cordova-plugin-appcenter-analytics" spec="^0.1.3" />
    <plugin name="cordova-plugin-appcenter-crashes" spec="^0.1.3" />
    <engine name="android" spec="~7.0.0" />

**Comment by BbFGE** [comment]
Update : Further investigation has shown there is an occasional problem installing the update
```
codePushUtil.js:41 [CodePush]  Cannot perform diff-update.. StackTrace: Error: Cannot perform diff-update.
    at file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/localPackage.js:358:37
    at FileReader.fileReader.onloadend (file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/fileUtil.js:232:17)
    at FileReader.readSuccessCallback (file:///android_asset/www/plugins/cordova-plugin-file/www/FileReader.js:164:18)
    at Object.callbackFromNative (file:///android_asset/www/cordova.js:291:58)
    at <anonymous>:1:9

```

ie from localpackage.js

```
  LocalPackage.handleDiffDeployment = function (newPackageLocation, diffManifest, diffCallback) {
        var handleError = function (e) {
            diffCallback(e, null);
        };
        LocalPackage.copyCurrentPackage(newPackageLocation, [".codepushrelease"], function (currentPackageError) {
            LocalPackage.handleCleanDeployment(newPackageLocation, function (cleanDeployError) {
                FileUtil.readFileEntry(diffManifest, function (error, content) {
                    if (error || currentPackageError || cleanDeployError) {
                        handleError(new Error("Cannot perform diff-update."));
```

Something is going wrong (on more than one machine). Is there a bug in the code perhaps?

**Comment by annakocheshkova** [comment]
Hi @BbFGE, this is a strange behavior. The logic of code push is that even after a rollback, a successfully released update must be installed. 

To tell you what's happening, we have to see the actual error. 
Could you modify the code of the `localpackage.ts` so that the error would be traced further, e.g. not `new Error("Cannot perform diff-update.")`, but `if (error) { new Error(error); }` etc.

To modify the code of the plugin, you can download it locally and run `gulp` from the root folder location after making the changes to the plugin code. 
You can install the local plugin via `cordova plugin add ../path/to/plugin`.

After you do that, please share with us the actual error that happens.

**Comment by BbFGE** [comment]
Hi
I tried the whole process from a different Windows machine and it seemed to allow multiple updates but then it started going wrong again and failing to update.
I attach a zip of the output for bloburl (not sure which one was right).

Note I am using appcenter commands not code-push eg
appcenter apps set-current myname/projectname
appcenter codepush release-cordova -d Staging

Hope you can help as if we can't reliably update using codepush, we will have to drop the whole project!

Thanks


[blob.zip](https://github.com/Microsoft/cordova-plugin-code-push/files/1664443/blob.zip)

**Comment by annakocheshkova** [comment]
@BbFGE I don't see anything strange in your package.

To tell you what's happening, we have to see the actual error. 
Could you modify the code of the `localpackage.ts` so that the error would be traced further, e.g. not `new Error("Cannot perform diff-update.")`, but `if (error) { handleError(error); }` etc for all the other errors.

To modify the code of the plugin, you can download it locally and run `gulp` from the root folder location after making the changes to the plugin code. 
You can install the local plugin via `cordova plugin add ../path/to/plugin`.

After you do that, please share with us the actual error that happens.

**Comment by BbFGE** [comment]
OK, I updated localpackage.ts

```
  FileUtil.readFileEntry(diffManifest, (error: Error, content: string) => {
                    if (error || currentPackageError || cleanDeployError) {
                        /*handleError(new Error("Tim Crannigan test Cannot perform diff-update."));*/
						if (error) { handleError(error); }
						if (currentPackageError) { handleError(currentPackageError); }
						if (cleanDeployError) { handleError(cleanDeployError); }
                    } else {
```

and implemented this in the app. The error is now 
```
[CodePush] Checking for update.
codePushUtil.js:36 [CodePush] An update is available. {"appVersion":"2.0.0","deploymentKey":"VkKh_e96R8TGJ_2x0OKYGGmeRIxrBk4hXPdrf","downloadUrl":"https://codepushupdates.azureedge.net/storagev2/CXb9JxmPRcPf0_OGgCu7HNYHm-4-22bdfca0-6228-41a3-83e0-14d8a18b2c42","isMandatory":false,"label":"v15","packageHash":"a40700c46553d6dd77fea0a99ab36f553a248696b9b63079e0f1a5d21f4dbf55","packageSize":714,"failedInstall":false}
codePushUtil.js:36 [CodePush] Downloading update
codePushUtil.js:36 [CodePush] Package download success: {"deploymentKey":"VkKh_e96R8TGJ_2x0OKYGGmeRIxrBk4hXPdrf","label":"v15","appVersion":"2.0.0","isMandatory":false,"packageHash":"a40700c46553d6dd77fea0a99ab36f553a248696b9b63079e0f1a5d21f4dbf55","isFirstRun":false,"failedInstall":false,"localPath":"cdvfile://localhost/files/codepush/download/update.zip"}
codePushUtil.js:36 [CodePush] Installing update
codePushUtil.js:41 [CodePush]  An error has occurred while performing the operation.  Error code: undefined. StackTrace: Error: An error has occurred while performing the operation.  Error code: undefined
    at Function.FileUtil.fileErrorToError (file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/fileUtil.js:22:16)
    at fail (file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/localPackage.js:336:63)
    at file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/codePushUtil.js:25:34
    at appDirError (file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/fileUtil.js:71:13)
    at fail (file:///android_asset/www/plugins/cordova-plugin-file/www/DirectoryEntry.js:77:9)
    at Object.callbackFromNative (file:///android_asset/www/cordova.js:293:52)
    at <anonymous>:1:9
CodePushUtil.logError @ codePushUtil.js:41
codePushUtil.js:41 [CodePush] An error occurred during sync. An error has occurred while performing the operation.  Error code: undefined. StackTrace: Error: An error has occurred while performing the operation.  Error code: undefined
    at Function.FileUtil.fileErrorToError (file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/fileUtil.js:22:16)
    at fail (file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/localPackage.js:336:63)
    at file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/codePushUtil.js:25:34
    at appDirError (file:///android_asset/www/plugins/cordova-plugin-code-push/bin/www/fileUtil.js:71:13)
    at fail (file:///android_asset/www/plugins/cordova-plugin-file/www/DirectoryEntry.js:77:9)
    at Object.callbackFromNative (file:///android_asset/www/cordova.js:293:52)
    at <anonymous>:1:9
```
It looks like some kind of problem with the downloaded file ie the new release.
If you have any ideas that would be great.

Thanks

**Comment by BbFGE** [comment]
I have done some more investigation and added some more code (temporarily) to the codepush plugin as suggested.

My test app is extremely simple : I set up a javascript variable holding a version number (not the actual version of the released app, just a number I can update to show on screen to prove a release has been updated).

I have then gone through cycles of editing the javascript and changing this variable, and using
`appcenter codepush release-cordova -d Staging`
to release.

This will work for a number of cycles and after each release, by debugging the device using Chrome, I can see the release has been installed and will run when the app is restarted 
![downloaded](https://user-images.githubusercontent.com/35726110/35513334-96796e84-04fa-11e8-82ad-cb54002bdf87.JPG)

I then restart the app, and it is ok.

However, after a certain number of cycles, it fails. I am fairly sure the log shows the release has been download but then when I restart, I think the app is unable to find the current version's directory (presumably it needs this to delete the previous release).
eg

![fail](https://user-images.githubusercontent.com/35726110/35513425-e1121aae-04fa-11e8-920d-a8270434712d.JPG)

I cannot work out what circumstances causes the fail : it can be after one or two cycles or many more. I have tried closing the app quickly after the app has downloaded the release to see if it is not updating a path correctly, but I can't see a pattern.

Once this error has occurred, there seems no way back on the device other than re-installing the whole app - even if I make new releases, it sticks on this error.

These are my changes to localpackage.js

```
  LocalPackage.copyCurrentPackage = function (newPackageLocation, ignoreList, copyCallback) {
        var handleError = function (e) {
            copyCallback && copyCallback(e, null);
        };
        var doCopy = function (currentPackagePath) {
            var getCurrentPackageDirectory;
            if (currentPackagePath) {
                getCurrentPackageDirectory = function (getCurrentPackageDirectoryCallback) {
                    console.log('getCurrentPackageDirectory executing. Next :  FileUtil.getDataDirectory. (currentPackagePath=' + currentPackagePath + ')');
                    FileUtil.getDataDirectory(currentPackagePath, false, getCurrentPackageDirectoryCallback);
                };
            }
            else {
                newPackageLocation = newPackageLocation + "/www";
                getCurrentPackageDirectory = function (getCurrentPackageDirectoryCallback) {
                    console.log('getCurrentPackageDirectory executing. Next :  FileUtil.getApplicationDirectory');
                    FileUtil.getApplicationDirectory("www", getCurrentPackageDirectoryCallback);
                };
            }
            FileUtil.getDataDirectory(newPackageLocation, true, function (deployDirError, deployDir) {
                if (deployDirError) {
                    handleError(new Error("Could not acquire the source/destination folders. "));
                }
                else {
                    var success = function (currentPackageDirectory) {
                        console.log('getCurrentPackageDirectory success ' + currentPackageDirectory);
                        FileUtil.copyDirectoryEntriesTo(currentPackageDirectory, deployDir, ignoreList, copyCallback);
                    };
                    var fail = function (fileSystemError) {
                        console.log('getCurrentPackageDirectory fail ' + fileSystemError);
                        copyCallback && copyCallback(FileUtil.fileErrorToError(fileSystemError), null);
                    };
                    console.log('Next : getCurrentPackageDirectory');
                    getCurrentPackageDirectory(CodePushUtil.getNodeStyleCallbackFor(success, fail));
                }
            });
        };
```

**Comment by BbFGE** [comment]
Have found more information but will raise a new issue

**Comment by 349989153** [comment]
Have no more details, just to mention that I ran into the same issue, still working on it.

After I release a new version, I manually checked for the update in my App, and then it downloaded, installed, restarted, and the new version showed up. However, if I turn it to background and then resume, its content change to apparently old version, if now I check for update manually, I get `SyncStatus.UP_TO_DATE`.

All I can confirm, is that `version` in the `config.xml` is always the same.So is the deployment is always `Staging`
